pkg/controller/v2/directive.go

```diff
@@ -0,0 +1,119 @@
+package controller
+
+import (
+	"math"
+	"time"
+)
+
+type Directive[I ID] interface {
+	Do(controller *Controller[I])
+}
+
+type Backoff func(attempt int) time.Duration
+
+func ConstantBackoff(delay time.Duration, maxDelay time.Duration) Backoff {
+	return func(attempt int) time.Duration {
+		return time.Duration(math.Min(float64(int(delay)*attempt), float64(maxDelay)))
+	}
+}
+
+func ExponentialBackoff(initialDelay time.Duration, maxDelay time.Duration) Backoff {
+	return func(attempt int) time.Duration {
+		maxExponent := math.Log2(float64(maxDelay) / float64(initialDelay))
+		return initialDelay * time.Duration(math.Pow(2, math.Min(float64(attempt), maxExponent)))
+	}
+}
+
+type Ack[I ID] struct {
+	request Request[I]
+}
+
+func (c *Ack[I]) Do(controller *Controller[I]) {
+	controller.Log.Debugf("Reconciliation of %s complete", c.request.ID)
+}
+
+type Requeue[I ID] struct {
+	request Request[I]
+}
+
+func (r *Requeue[I]) Do(controller *Controller[I]) {
+	controller.Log.Debugf("Requeueing %s", r.request.ID)
+	go controller.enqueue(r.request)
+}
+
+type Fail[I ID] struct {
+	request Request[I]
+	Error   error
+}
+
+func (f *Fail[I]) Do(controller *Controller[I]) {
+	controller.Log.Warnf("Reconciliation of %s failed: %s", f.request.ID, f.Error.Error())
```

Comment:
should we change these logs to use Warnw and make it structured?

Reply:
done